File: python/2022/03/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def val(a):
    if ord(a) >= ord('a') and ord(a) <= ord('z'):
        return ord(a) - ord('a') + 1
    if ord(a) >= ord('A') and ord(a) <= ord('Z'):
        return ord(a) - ord('A') + 27
    raise Exception('Unknown letter')

def comp(a,b):
    for aa in a:
        if aa in b:
            return val(aa)

def find(a,b,c):
    x = []
    for aa in a:
        if aa in b:
            x.append(aa)
    for xx in x:
        if xx in c:
            return val(xx)

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    B = [comp(a[:len(a)//2], a[len(a)//2:]) for a in A]
    k = sum(B)
    print("Part 1:",k)

def part2():
    A = read_input()
    B = [find(A[3*i], A[3*i+1], A[3*i+2]) for i in range(len(A)//3)]
    k = sum(B)
    print("Part 2:",k)

part1()
part2()

File: pascal/2022/03/main.pas
program main;
uses 
	util;

const
	debug = false;

	sampleFile = 'sample.txt';
	inputFile = 'input.txt';	

var
	fp: string;
	f: text;
	
	function eval(c: char): integer;
	begin
		if (ord(c) >= ord('a')) and (ord(c) <= ord('z')) then
			eval := ord(c) - ord('a') + 1
		else eval := ord(c) - ord('A') + 27;
	end;
	
	procedure part1;
	var 
		A: array of string;
		i, l, n: integer;	
		aa, q, w: string;
	begin		
		A := util.readInput(f);
		i := 0;
		for aa in A do
		begin
			l := round(length(aa)/2);
			q := copy(aa, 0, l);
			w := copy(aa, l+1, l); 
			for n := 1 to l+1 do
			begin
				if pos(q[n], w) > 0 then
				begin
					i := i + eval(q[n]);
					break;
				end;
			end;
		end;
		WriteLn('Part 1: ',i);
	end;

	procedure part2;
	var 
		A: array of string;
		i, j, l, n: integer;	
		q: string;
	begin		
		A := util.readInput(f);
		i := 0;
		for j := 0 to (length(A) div 3)-1 do begin
			q := A[3*j];
			l := round(length(q));
			for n := 1 to l+1 do
			begin
				if (pos(q[n], A[3*j+1]) > 0) and (pos(q[n], A[3*j+2]) > 0) then
				begin
					i := i + eval(q[n]);
					break;
				end;
			end;
		end;
		WriteLn('Part 2: ',i);
	end;
begin
	if debug then
		fp := sampleFile
	else
		fp := inputFile;	
	assign(f, fp);

	part1;
	part2;

end.
